File: S4-Clients/Java-client/src/main/java/com/ontotext/s4/service/impl/S4AbstractClientImpl.java
package com.ontotext.s4.service.impl;


import com.fasterxml.jackson.databind.JsonNode;
import com.ontotext.s4.catalog.ServiceDescriptor;
import com.ontotext.s4.client.HttpClient;
import com.ontotext.s4.client.HttpClientException;
import com.ontotext.s4.service.S4AbstractClient;
import com.ontotext.s4.service.util.S4ServiceClientException;

import java.net.MalformedURLException;
import java.net.URL;

public class S4AbstractClientImpl implements S4AbstractClient {

    protected HttpClient client;
    protected boolean requestCompression;

    private S4AbstractClientImpl() {}
    
    public S4AbstractClientImpl(ServiceDescriptor item, String apiKeyId, String apiPassword) {
        URL endpoint;
        try {
            endpoint = new URL(item.getServiceUrl());
        } catch (MalformedURLException murle) {
            throw new IllegalArgumentException("Invalid ServiceDescriptor specified. No API endpoint found.", murle);
        }
        this.client = new HttpClient(endpoint, apiKeyId, apiPassword);
    }

    public S4AbstractClientImpl(URL endpoint, String apiKeyId, String apiPassword) {
        this.client = new HttpClient(endpoint, apiKeyId, apiPassword);
    }

    public void setRequestCompression(boolean requestCompression) {
        this.requestCompression = requestCompression;
    }

    /**
     * Processes a specific {@link HttpClientException} and returns its message.
     * *NOTE* This is a utility helper method
     * @param e An {@link HttpClientException} to process
     * @return The error message from the exception
     */
    protected JsonNode handleErrors(HttpClientException e) {
        JsonNode response = e.getResponse();
        if(response == null) {
            throw new S4ServiceClientException(e.getMessage(), e);
        }
        return response.get("message");
    }
}

File: S4-Clients/C#-client/S4ClientApi/Service/Impl/S4AbstractClientImpl.cs
/*
 * S4 C# client library
 * Copyright 2016 Ontotext AD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using Ontotext.S4.catalog;
using Ontotext.S4.client;
using Ontotext.S4.common;
using Ontotext.S4.service;
using System;
using System.Collections.Generic;
using System.Json;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S4ClientApi.Service.Impl
{
    public class S4AbstractClientImpl : IS4AbstractClientImpl
    {
        protected HttpClient client;
        protected bool RequestCompression;

        /// <summary>
        /// Construct a <code>S4ServiceClient</code> accessing a specific processing pipeline on the s4.ontotext.com platform using the given credentials.
        /// </summary>
        /// <param name="item">the {@link ServiceDescriptor} which represents the processing pipeline which will be used</param>
        /// <param name="apiKeyId">API key ID for authentication</param>
        /// <param name="keySecret">corresponding password</param>
        public S4AbstractClientImpl(ServiceDescriptor item, String apiKeyId, String keySecret)
        {
            Uri endpoint;
            try
            {
                endpoint = new Uri(item.ServiceUrl);
            }
            catch (Exception e)
            {
                throw new ArgumentException("Invalid ServiceDescriptor specified. No API endpoint specified." + e.Message);
            }
            this.client = new HttpClient(endpoint, apiKeyId, keySecret);
        }

        /// <summary>
        /// Construct a <code>S4ServiceClient</code> accessing a specific processing pipeline on the s4.ontotext.com platform using the given credentials.
        /// </summary>
        /// <param name="endpoint">the URL of the pipeline which will be used for processing</param>
        /// <param name="apiKeyId">apiKeyId API key ID for authentication</param>
        /// <param name="keySecret">corresponding password</param>
        public S4AbstractClientImpl(Uri endpoint, String apiKeyId, String keySecret)
        {
            this.client = new HttpClient(endpoint, apiKeyId, keySecret);
        }

        public bool requestCompression
        {
            get { return RequestCompression; }
            set { RequestCompression = value; }
        }

        protected String handleErrors(HttpClientException e)
        {
            JsonObject response = e.getResponse();
            if (response == null)
            {
                throw new S4ServiceClientException(e.Message, e);
            }
            return response.ToString();
        }

        protected Dictionary<String, String> constructHeaders(ResponseFormat serializationFormat)
        {
            Dictionary<String, String> headers = Parameters.newInstance().withValue("Accept", serializationFormat.ToStringValue());
            if (requestCompression)
            {
                headers.Add("Accept-Encoding", "gzip");
            }
            return headers;
        }
    }
}
